core/blockchain.go

```diff
@@ -788,19 +788,21 @@ func (self *BlockChain) WriteBlock(block *types.Block) (status WriteStatus, err
 				return NonStatTy, err
 			}
 		}
-		// Insert the block as the new head of the chain
-		self.insert(block)
 		status = CanonStatTy
 	} else {
 		status = SideStatTy
 	}
+
 	// Irrelevant of the canonical status, write the block itself to the database
 	if err := self.hc.WriteTd(block.Hash(), block.NumberU64(), externTd); err != nil {
```

Comment:
Same here actually. The change would be simpler if the WriteBlock, WriteTd calls were simply moved up instead of moving the self.insert call after them.